### ISSUE TITLE
[Feature] Dataset API - Reinitializable Iterator resets to first dataset element

### ISSUE THREAD

**Issue by maxfiedler** [tensorflow/tensorflow#15448]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: slightly altered stock example (see below)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: 1.4.0 from source
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: 0.7.0
- **GCC/Compiler version (if compiling from source)**: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
- **CUDA/cuDNN version**: 8.0
- **GPU model and memory**: Nvidia 1060

### Describe the problem
Currently the re-initializable iterator API using .from_structure and Iterator.make_initializer always resets the get_next() op of the iterator to fetch the first element of its Dataset instance again after running `sess.run(training_init_op)` or `sess.run(validation_init_op)`, respectively. 

Is this really the intended behavior? This means, that if you want to switch between training and validation Datasets within one epoch (i.e. in a shorter rhythm than the full dataset length) you will always only iterate over the first `batch_size * number-of-training-steps-before-validation-step` elements of the training set during training. 

### Source code / logs
I guess the code example will make it clearer:
```
# Define training and validation datasets with the same structure.
training_dataset = tf.data.Dataset.range(100)
validation_dataset = tf.data.Dataset.range(50)

# A reinitializable iterator is defined by its structure. We could use the
# `output_types` and `output_shapes` properties of either `training_dataset`
# or `validation_dataset` here, because they are compatible.
iterator = tf.data.Iterator.from_structure(training_dataset.output_types,
                                  training_dataset.output_shapes)

next_element = iterator.get_next()


training_init_op = iterator.make_initializer(training_dataset)
validation_init_op = iterator.make_initializer(validation_dataset)

with tf.Session() as sess:
   # Run 20 epochs in which the training dataset is traversed, followed by the
   # validation dataset.
   for i in range(5):
       # Initialize an iterator over the training dataset.
       print("#########################  ", i)
       sess.run(training_init_op)
       for _ in range(10):
           nel = sess.run(next_element)
           print("train: ", type(nel), nel)

       # Initialize an iterator over the validation dataset.
       sess.run(validation_init_op)
       for _ in range(5):
           nel = sess.run(next_element)
           print("valid: ", type(nel), nel)
```

Produces the output:

```
#########################   0
train:  <class 'numpy.int64'> 0
train:  <class 'numpy.int64'> 1
train:  <class 'numpy.int64'> 2
train:  <class 'numpy.int64'> 3
train:  <class 'numpy.int64'> 4
train:  <class 'numpy.int64'> 5
train:  <class 'numpy.int64'> 6
train:  <class 'numpy.int64'> 7
train:  <class 'numpy.int64'> 8
train:  <class 'numpy.int64'> 9
valid:  <class 'numpy.int64'> 0
valid:  <class 'numpy.int64'> 1
valid:  <class 'numpy.int64'> 2
valid:  <class 'numpy.int64'> 3
valid:  <class 'numpy.int64'> 4
#########################   1
train:  <class 'numpy.int64'> 0
train:  <class 'numpy.int64'> 1
train:  <class 'numpy.int64'> 2
train:  <class 'numpy.int64'> 3
train:  <class 'numpy.int64'> 4
train:  <class 'numpy.int64'> 5
train:  <class 'numpy.int64'> 6
train:  <class 'numpy.int64'> 7
train:  <class 'numpy.int64'> 8
train:  <class 'numpy.int64'> 9
valid:  <class 'numpy.int64'> 0
valid:  <class 'numpy.int64'> 1
valid:  <class 'numpy.int64'> 2
valid:  <class 'numpy.int64'> 3
valid:  <class 'numpy.int64'> 4
...
```
Apparently, the latter 90 elements of the training set and the latter 45 elements of the validation set never get evaluated. I don't really see the real-world use-case for this behavior. 

I know that you can implement the other functionality via the feedable iterator scheme using one_shot_iterators (but not using initializable iterators) as highlighted by the code below (pay attention to the different iterators used for training and validation here):

```
# Define training and validation datasets with the same structure.
training_dataset = tf.data.Dataset.range(10000000).repeat(2)
validation_dataset = tf.data.Dataset.range(5000000).repeat(2)

# A feedable iterator is defined by a handle placeholder and its structure. We
# could use the `output_types` and `output_shapes` properties of either
# `training_dataset` or `validation_dataset` here, because they have
# identical structure.
handle = tf.placeholder(tf.string, shape=[])
iterator = tf.data.Iterator.from_string_handle(
    handle, training_dataset.output_types, training_dataset.output_shapes)
next_element = iterator.get_next()

# You can use feedable iterators with a variety of different kinds of iterator
training_iterator = training_dataset.make_one_shot_iterator()
validation_iterator = validation_dataset.make_initializable_iterator()

with tf.Session() as sess:
    # The `Iterator.string_handle()` method returns a tensor that can be evaluated
    # and used to feed the `handle` placeholder.
    training_handle = sess.run(training_iterator.string_handle())
    validation_handle = sess.run(validation_iterator.string_handle())
    # Loop forever, alternating between training and validation.
    for i in range(5):
        print("######################## ", i)
        i += 1
        # Run 10 steps using the training dataset. Note that the training dataset is
        # 2 * the original set, i.e. we run 2 epochs (see .repeat() argument), and we resume from where
        # we left off in the previous `while` loop iteration.
        for _ in range(10):
            nel = sess.run(next_element, feed_dict={handle: training_handle})
            print("train: ", type(nel), nel)

        # Run one pass over the validation dataset.
        sess.run(validation_iterator.initializer)
        for _ in range(5):
            nel = sess.run(next_element, feed_dict={handle: validation_handle})
            print("valid: ", type(nel), nel)
```

creates output:

```
########################  0
train:  <class 'numpy.int64'> 0
train:  <class 'numpy.int64'> 1
train:  <class 'numpy.int64'> 2
train:  <class 'numpy.int64'> 3
train:  <class 'numpy.int64'> 4
train:  <class 'numpy.int64'> 5
train:  <class 'numpy.int64'> 6
train:  <class 'numpy.int64'> 7
train:  <class 'numpy.int64'> 8
train:  <class 'numpy.int64'> 9
valid:  <class 'numpy.int64'> 0
valid:  <class 'numpy.int64'> 1
valid:  <class 'numpy.int64'> 2
valid:  <class 'numpy.int64'> 3
valid:  <class 'numpy.int64'> 4
########################  1
train:  <class 'numpy.int64'> 10
train:  <class 'numpy.int64'> 11
train:  <class 'numpy.int64'> 12
train:  <class 'numpy.int64'> 13
train:  <class 'numpy.int64'> 14
train:  <class 'numpy.int64'> 15
train:  <class 'numpy.int64'> 16
train:  <class 'numpy.int64'> 17
train:  <class 'numpy.int64'> 18
train:  <class 'numpy.int64'> 19
valid:  <class 'numpy.int64'> 0
valid:  <class 'numpy.int64'> 1
valid:  <class 'numpy.int64'> 2
valid:  <class 'numpy.int64'> 3
valid:  <class 'numpy.int64'> 4
########################  2
train:  <class 'numpy.int64'> 20
train:  <class 'numpy.int64'> 21
train:  <class 'numpy.int64'> 22
train:  <class 'numpy.int64'> 23
train:  <class 'numpy.int64'> 24
train:  <class 'numpy.int64'> 25
train:  <class 'numpy.int64'> 26
train:  <class 'numpy.int64'> 27
train:  <class 'numpy.int64'> 28
train:  <class 'numpy.int64'> 29
valid:  <class 'numpy.int64'> 0
valid:  <class 'numpy.int64'> 1
valid:  <class 'numpy.int64'> 2
valid:  <class 'numpy.int64'> 3
valid:  <class 'numpy.int64'> 4
```

**Comment by nikste** [comment]
I'm having the same problem, how can you switch between training and validation pipeline, without resetting the iterators?

**Comment by kmhofmann** [comment]
Indeed, I have also observed the same issue.
Prior to using the Dataset API, I could implement multiple queues that would fetch data in parallel; what was unfortunately not possible was to reset (reuse) them after an epoch, e.g. for repeated evaluation of the validation set.

Now, with a re-initializable iterator, the queue(s) seem to be coupled to the iterator and will be forcibly reset by switching (if I see this correctly), which makes the whole thing a bit pointless to be used with training data; see OP's example.

I also cannot see good use cases for this behavior. In our code base, we intend to evaluate the validation set periodically after a certain number of training iterations, multiple times per training epoch. For this (supposedly quite common) use case seamless switching is necessary.

**Comment by mrry** [comment]
It sounds like the reinitializable iterator is working as intended (resetting to its initial state when you run the initializer). You should use the feedable iterator if you want to switch between two iterators without resetting. Feedable iterators work with both one-shot and initializable iterators. Feel free to reopen the bug if the feedable iterators aren't working as expected.

**Comment by kmhofmann** [comment]
IIRC there's a performance hit when using the feedable API. Is that still the case or not?

**Comment by mrry** [comment]
I haven't seen any specific reports about a performance hit. Feeding an extra value might cost ~10s of microseconds per step, but that's usually in the noise.

**Comment by maxfiedler** [comment]
@mrry Thanks for clarifying. I am still wondering what real-life use case the re-initializable iterators serve then? To me it seems that it can only be used practically for complete batch optimization but not for stochastic gradient descent or mini-batch or if I am content to get a validation loss only after each full training epoch. 
Or are there other possibilities to manipulate your dataset in such a way that you can still intertwine training and validation set mini batches using the re-initializable iterator API?

**Comment by mrry** [comment]
Reinitializable iterators allow you to change the parameters of a dataset by feeding placeholders when you initialize them. This lets you do things like changing the training or test set at the end of an epoch to reflect new input files that have been added in the meantime. Even if they aren't appropriate for your use case, remember that TensorFlow users are a diverse and creative bunch, and might have different requirements!

**Comment by maxfiedler** [comment]
I see, thanks for laying it out to me :)

**Comment by baodingge** [comment]
I'm having the same problem，I train my network always using the process:
for i in range(train_step):
    sess.run(train_op_with_next_train_batch)
    if i%100==0:
        sess.run(test_with_next_test_batch)
but, i can't do that with tf.data, .make_one_shot_iterator() can only train data one epoch, make_initializable_iterator() will restart from the begining, how to solve this problem?
I try to solve this problem by set data_set.repead(10000), but i don't think this is a good ieda.
Do you have any idea?thanks very much.

**Comment by matwilso** [comment]
It looks like you could manually [save and restore the state of the iterator](https://www.tensorflow.org/programmers_guide/datasets#saving_iterator_state), and this would allow you to swap between different datasets without resetting to the first dataset element.  I haven't tried it yet, though, so I can't guarantee for sure.
